source/module_elecstate/module_dm/density_matrix.cpp

```diff
@@ -24,62 +24,24 @@ DensityMatrix<TK, TR>::~DensityMatrix()
     }
 }
 
-// constructor for multi-k
 template <typename TK, typename TR>
-DensityMatrix<TK, TR>::DensityMatrix(const K_Vectors* kv_in, const Parallel_Orbitals* paraV_in, const int nspin)
+DensityMatrix<TK, TR>::DensityMatrix(const Parallel_Orbitals* paraV_in, const int nspin, const std::vector<ModuleBase::Vector3<double>>& kvec_d, const int nk)
+    : _paraV(paraV_in), _nspin(nspin), _kvec_d(kvec_d), _nk((nk > 0 && nk <= _kvec_d.size()) ? nk : _kvec_d.size())
 {
     ModuleBase::TITLE("DensityMatrix", "DensityMatrix-MK");
-    this->_kv = kv_in;
-    this->_paraV = paraV_in;
-    // set this->_nspin
-    if (nspin == 1 || nspin == 4)
-    {
-        this->_nspin = 1;
-    }
-    else if (nspin == 2)
-    {
-        this->_nspin = 2;
-#ifdef __DEBUG
-        assert(kv_in->get_nks() % 2 == 0);
-#endif
-    }
-    else
-    {
-        throw std::string("nspin must be 1, 2 or 4");
-    }
-    // set this->_nks, which is real number of k-points
-    this->_nks = kv_in->get_nks() / this->_nspin;
-    // allocate memory for _DMK
-    this->_DMK.resize(this->_kv->get_nks());
-    for (int ik = 0; ik < this->_kv->get_nks(); ik++)
+    const int nks = _nk * _nspin;
```

Comment:
If one set `nk=kv.nks` & `nspin=2`. Here `nks` will be `2*kv.nks`. How to avoid this? We cannot assume everyone knows exactly what they do.

Reply:
I think this is an ill-design of class `K_Vectors`. `DensityMatrix` itself does not depend on it any more.
For a general `DensityMatrix` object, this is allowd as long as `kv.nks <= kv.kvec_d[ik].size()`. The object can do Fourier transformation normally, but costs twice as much memory (R->k) or gets a doubled result (k->R) due to the repeat of k-points in `kv.kvec_d`.

Reply:
The `nks` here need not to be equal to `kv.nks`. For example, in LR-TDDFT I will pass real nk and nspin=1 to `DensityMatrix` when global nspin=2, the `nks` is half of `kv.nks`, which is exactly what I want to do.

Reply:
If one construct a DensityMatrix by:
```cpp
DensityMatrix<TK, TR>::DensityMatrix(paraV_in, 2, kvec_d);
```
The size of DMK is:
`DMK.size()=2*kvec_d.size()`
However, in old code, `DMK.size()=kvec_d.size()`. Why do we need this extra allocation?

Reply:
I think the size of `K_Vectors::kvec_d` should be shrink to `nk` rather than `nks=nk*nspin`. It's already a redundance when nspin=2 -- still a design problem of class `K_Vectors`.

Reply:
I think the size of `K_Vectors::kvec_d` should be shrink to `nk` rather than `nks=nk*nspin`, because the coordinates of k-points are independent of spin. It's already a redundance when nspin=2 -- still a design problem of class `K_Vectors`...
Once shrinked so, `DMK.size()=2*kvec_d.size()` is resonable at nspin=2.

Reply:
I recognize your point of view about 

> the size of K_Vectors::kvec_d should be shrinked to nk rather than nks=nk*nspin.

 However, I lean towards the idea that development should be based on the existing code design, rather than an ideal future design. We still need to consider the current state of `kvec_d` when modifying our `DMK`. As for the redesign and refactoring of `kvec_d`, that should be part of a new PR. I don't think it's a good idea to default to allocating twice the DMK for all calculations with nspin=2 in the construction function of DM; I'm concerned about the potential risks it might introduce during program execution that we cannot predict. 

I understand that you want to support the needs you're currently facing within openshell-LR-TDDFT but is this the way to go about it? Due to my personal level of expertise, I am unable to foresee the potential risks of doing so. Does anyone else have any suggestions? I welcome further discussion. @mohanchen @dyzheng @jinzx10 @kirk0830